File: moamoa_front/src/model/ApiResponse.ts
export interface ApiResponse<T> {
    code: string;
    message: string;
    data: T | null;
}

export interface ApiSuccessResponse<T> extends ApiResponse<T> {
    code: string;
    message: string;
    data: T;
}

export interface ApiErrorResponse extends ApiResponse<null> {
    code: string;
    message: string;
    data: null;
}

export const isApiSuccess = <T>(response: ApiResponse<T>): response is ApiSuccessResponse<T> => {
    return response.data !== null;
};

export const isApiError = <T>(response: ApiResponse<T>): response is ApiErrorResponse => {
    return response.code !== "" && response.data === null;
};

File: src/main/java/be/geon/moa/moamoa/ApiResponse.java
package be.geon.moa.moamoa;

import be.geon.moa.moamoa.config.ErrorCode;
import lombok.*;


@ToString
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class ApiResponse<T> {
    public static final ApiResponse<Integer> SUCCESS_WITH_INT_ONE = success(1);
    public static final ApiResponse<Integer> SUCCESS_WITH_INT_ZERO = success(0);

    private String code;
    private String message;
    private T data;

    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>("", "", data);
    }

    public static <T> ApiResponse<Integer> maskToInteger(T data) {
        return data != null ? SUCCESS_WITH_INT_ONE : SUCCESS_WITH_INT_ZERO;
    }

    public static ApiResponse<Object> error(ErrorCode errorCode) {
        return new ApiResponse<>(errorCode.getCode(), errorCode.getMessage(), null);
    }

    public static ApiResponse<Object> error(ErrorCode errorCode, String message) {
        return new ApiResponse<>(errorCode.getCode(), message, null);
    }
}